move legacy UnifiedJob stdout data to a separate unmanaged model

ryanpetrello opened the pull request:
This data often (in the case of inventory updates) represents large data
blobs (5+MB per job run).  Storing it on the polymorphic base class
table, `main_unifiedjob`, causes it to be automatically fetched on every
query (and every polymorphic join) against that table, which can result
in _very_ poor performance for awx across the board.  Django offers
`defer()`, but it's quite complicated to sprinkle this everywhere (and
easy to get wrong/introduce side effects related to our RBAC and usage
of polymorphism).

This change moves the field definition to a separate unmanaged model
(which references the same underlying `main_unifiedjob` table) and adds
a proxy for fetching the data as needed

see #200

ryanpetrello commented on awx/main/migrations/0012_move_deprecated_stdout.py:
```diff
@@ -0,0 +1,33 @@
+            name='job',
+            field=models.OneToOneField(editable=False, null=True, on_delete=django.db.models.deletion.CASCADE, related_name='deprecated_result_stdout', to='main.UnifiedJob'),
+        ),
+        migrations.RunSQL('INSERT INTO main_deprecatedjobstdout(job_id, result_stdout_text) SELECT id, result_stdout_text FROM main_unifiedjob WHERE result_stdout_text IS NOT NULL;'),
```
I've got a feeling @matburt ain't gonna like this :)

That said, I _really_ feel like we should avoid doing this in Python - it's just not going to be fast (or memory-performant) enough.

ryanpetrello replied:
I'll also add - I'm not sure how long it actually would take *postgres* to do this for really large datasets - I plan to load a GB or so of stdout and attempt this query.

ryanpetrello commented on awx/main/models/__init__.py:
```diff
@@ -145,17 +145,3 @@ def user_is_in_enterprise_category(user, category):
-
-# Always, always, always defer result_stdout_text for polymorphic UnifiedJob rows
-# TODO: remove this defer in 3.3 when we implement https://github.com/ansible/ansible-tower/issues/5436
-def defer_stdout(f):
```
Holy cow, it feels really good to get rid of this.

ryanpetrello commented on awx/main/models/unified_jobs.py:
```diff
@@ -882,6 +877,22 @@ def create_config_from_prompts(self, kwargs):
+            return ''
+
+    @result_stdout_text.setter
+    def result_stdout_text(self, value):
```
This setter will need to continue to exist for the time being; my eventual goal once https://github.com/ansible/awx/issues/200 is fully implemented is for this method to totally go away (because no _new_ job runs should be storing stdout in this table).

AlanCoding replied:
Might be good to mark the method for future deletion with a TODO?

AlanCoding commented on awx/main/migrations/0012_move_deprecated_stdout.py:
```diff
@@ -0,0 +1,33 @@
+            model_name='deprecatedjobstdout',
+            name='job',
+            field=models.OneToOneField(editable=False, null=True, on_delete=django.db.models.deletion.CASCADE, related_name='deprecated_result_stdout', to='main.UnifiedJob'),
+        ),
```
I don't see why there should be 2 separate operations here. Could you roll the `deprecatedjobstdout` field creation into the `CreateModel` by deleting these and running `makemigrations` again?

ryanpetrello replied:
Yea, I was confused by this, too, but it's what `makemigrations` created :/

AlanCoding replied:
I checked out your branch, and that method gave me

```python
# -*- coding: utf-8 -*-
# Generated by Django 1.11.7 on 2017-12-12 14:44
from __future__ import unicode_literals

from django.db import migrations, models
import django.db.models.deletion


class Migration(migrations.Migration):

    dependencies = [
        ('main', '0013_auto_20171212_1444'),
    ]

    operations = [
        migrations.CreateModel(
            name='DeprecatedJobStdout',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('result_stdout_text', models.TextField(blank=True, default=b'', editable=False)),
                ('job', models.OneToOneField(editable=False, null=True, on_delete=django.db.models.deletion.CASCADE, related_name='deprecated_result_stdout', to='main.UnifiedJob')),
            ],
        ),
    ]
```

ryanpetrello replied:
That's so bizarre O_O, let me blow away my database and give this another look.

ryanpetrello commented on awx/main/migrations/0012_move_deprecated_stdout.py:
```diff
@@ -0,0 +1,29 @@
+                ('job', models.OneToOneField(editable=False, null=True, on_delete=django.db.models.deletion.CASCADE, related_name='deprecated_result_stdout', to='main.UnifiedJob'))
+            ],
+        ),
+        migrations.RunSQL('INSERT INTO main_deprecatedjobstdout(job_id, result_stdout_text) SELECT id, result_stdout_text FROM main_unifiedjob WHERE result_stdout_text IS NOT NULL;'),
```
I've got a feeling @matburt ain't gonna like this :)

That said, I really feel like we should avoid doing this in Python - it's just not going to be fast (or memory-performant) enough.  I'm not sure how long it actually would take postgres to do this for really large datasets - I plan to load a GB or so of stdout and attempt this query.

AlanCoding reviewed (approved):
I approve everything except the SQL code, which I leave for @matburt

ryanpetrello commented on awx/main/migrations/0012_move_deprecated_stdout.py:
```diff
@@ -0,0 +1,29 @@
+                ('job', models.OneToOneField(editable=False, null=True, on_delete=django.db.models.deletion.CASCADE, related_name='deprecated_result_stdout', to='main.UnifiedJob'))
+            ],
+        ),
+        migrations.RunSQL("INSERT INTO main_deprecatedjobstdout(job_id, result_stdout_text) SELECT id, result_stdout_text FROM main_unifiedjob WHERE result_stdout_text != '';"),
```
I'm testing and I'm still on the fence about whether or not this approach is going to be performant enough for very large datasets.  Still experimenting...

wenottingham replied:
As a one-time migration of legacy data, the bar for "performant enough" (assuming no OOM issues) is pretty low.

ryanpetrello committed:
move legacy UnifiedJob stdout data to a separate unmanaged model

This data often (in the case of inventory updates) represents large data
blobs (5+MB per job run).  Storing it on the polymorphic base class
table, `main_unifiedjob`, causes it to be automatically fetched on every
query (and every polymorphic join) against that table, which can result
in _very_ poor performance for awx across the board.  Django offers
`defer()`, but it's quite complicated to sprinkle this everywhere (and
easy to get wrong/introduce side effects related to our RBAC and usage
of polymorphism).

This change moves the field definition to a separate unmanaged model
(which references the same underlying `main_unifiedjob` table) and adds
a proxy for fetching the data as needed

see https://github.com/ansible/awx/issues/200

ryanpetrello commented:
@matburt @AlanCoding @chrismeyersfsu 

I changed my approach so that we're not actually copying data around (which is _really_ time consuming for large data sets).  Instead, the data *stays* in `main_unifiedjob.result_stdout_text`, but that field is removed from the `UnifiedJob` class, so it won't be fetched in ORM queries.

For _direct_ access to the column, I've added a property that proxies the `result_stdout_text` column through a new virtual, unmanaged model:

https://github.com/ansible/awx/pull/795/files#diff-160977429605d785dda624eb08043a88R892

I've still got to fix up some tests, but this mostly seems to work.

ryanpetrello commented on awx/main/models/unified_jobs.py:
```diff
@@ -492,6 +492,18 @@ def model_to_str(self):
+class UnifiedJobDeprecatedStdout(models.Model):
+
+    class Meta:
+        managed = False
```
The key to this change is here:

1.  Remove `result_stdout_text` from the `UnifiedJob` class definition so it's no longer a field that gets tacked onto every query that the ORM builds.
2.  Introduce a migration that allows the field to be removed from `UnifiedJob` _without_ dropping the underlying column (so all of the legacy data still exists post-migration): https://github.com/ansible/awx/pull/795/files#diff-0d54c0f6b7264d1e5a102d50f1417810R23
3.  Add the `result_stdout_text` field to this _new_ model, which is marked as `managed = False` and `db_table = 'main_unifiedjob'`.  This means that this model will pull its field data from the `main_unifiedjob` instead of creating a new table for this model; this approach is what you do when you want to manage database views in Django, and if you think about this, it's sort of the same thing.
4.  Add a getter and setter to the original `UnifiedJob` class so that direct access is still available (which proxies through to this new underlying model):
https://github.com/ansible/awx/pull/795/files#diff-160977429605d785dda624eb08043a88R892

matburt replied:
👍

matburt commented on awx/main/migrations/0012_move_deprecated_stdout.py:
```diff
@@ -0,0 +1,44 @@
+            name='result_stdout_text',
+            field=models.TextField(editable=False, null=True),
+        ),
+        # Using SeparateDatabaseAndState here allows us to update the migration
```
![til](https://user-images.githubusercontent.com/89544/33905979-7b1c15e6-df4e-11e7-9256-f660f29f9dd9.gif)

matburt reviewed (approved):
lgtm

ryanpetrello commented on awx/main/tests/functional/__init__.py:
```diff
@@ -0,0 +1,23 @@
+from django.apps import apps
+
+
+def app_post_migration(sender, app_config, **kwargs):
```
@matburt @AlanCoding @chrismeyersfsu 

This is a somewhat gross hack I had to add to make pytest-django + sqlite happy.  The issue is that for *real* installs and upgrades, we run `awx-manage migrate`, which causes `main_unifiedjob.result_stdout_text` to exist under the hood _despite_ its absence from the `UnifiedJob` model definition (due to https://github.com/ansible/awx/pull/795/files#r156477183).

Our functional tests work differently.  They set up a sqlite database created by `pytest-django`.  This process explicitly _does not_ run migrations (because it's slow), but instead creates the database schema using the _current state_ of the model definitions.  The result is that the schema spun up by `pytest-django` doesn't have this column (because it no longer exists on `UnifiedJob`).  This migration hook is a bit of a kludge, but it _does_ give us what we want - a schema for tests that matches reality.

ryanpetrello commented on awx/main/tests/unit/test_unified_jobs.py:
```diff
@@ -8,15 +8,15 @@
 # stdout file present
 @mock.patch('os.path.exists', return_value=True)
 @mock.patch('codecs.open', return_value='my_file_handler')
+@mock.patch.object(models.UnifiedJob, 'result_stdout_text', '')
```
These unit tests have changed to just mock the `result_stdout_text` property; they don't need to care about the underlying implementation details of how the data gets stored and retrieved.

ryanpetrello commented:
I wish Django's ORM had better tools for this sort of problem.

The end goal of of my PR is that we'll no longer have to use `defer()` everywhere (and so we don't have to worry about forgetting it in other places in the future, thus introducing new performance bugs when we forget it).

Also, at some point in the future, we'll eventually arrive at a place where I think it will be appropriate to _entirely lose_ this column because it represents years-old legacy stdout data.  For now, this gives us some of the foundational bits we need to accomplish https://github.com/ansible/awx/issues/200 in a backwards-compatible way.
